### PR TITLE
[FEATURE] DBAL 3

### PR DESCRIPTION
**Pull request by keithbrink**
Following on the work of #519 and #520, I went through the code to fix all the breaking changes related to DBAL 3, as well as some deprecations that came up.

I would prefer to target a new 2.0 branch with this, since there are many breaking changes.

- Removed the Json type since the `json_array` type has been removed in DBAL 3.
- Removed the MasterSlaveConnection as it has been removed from DBAL 3.
- Removed the FluentExporter, GenerateEntitiesCommand, GenerateRepositoriesCommand, ConvertMappingCommand, as in #520
- Removed the flush option from the various cache clearing commands; these have been in place since the initial commit to this repository, so I couldn't find any reason why it existed in the first place. (Ideas? @patrickbrouwers) Flush is no longer supported by Symfony cache anyways.
- Remove the deprecated namespace feature.
- Updated the PHP CS Fixer file to the new format, using the PSR12 ruleset. Can move this to a separate PR if preferred.
- Bumped the doctrine/orm version requirement to 2.14 due to the new ORMSetup class and the deprecated `create` method on the EntityManager: https://github.com/doctrine/orm/pull/9961

### CODE REVIEW

**Comment by dpslwk** [comment]
Any reason you did no start with #520 ??
We were close on that PR and just a few tweaks requested, was goin to pick it up myself later in the month to finish off

**Comment by keithbrink** [comment]
@dpslwk Didn't see it! I'll have a look at it now.

**Comment by keithbrink** [comment]
Working on converting the caches over, marked as draft for now.

**Comment by dpslwk** [comment]
`MasterSlaveConnection` was not remove so much as changed to `PrimaryReadReplicaConnection`
we had a request to make the same changes https://github.com/laravel-doctrine/orm/pull/520#discussion_r973195088 in #520

**Comment by keithbrink** [comment]
> `MasterSlaveConnection` was not remove so much as changed to `PrimaryReadReplicaConnection` we had a request to make the same changes [#520 (comment)](https://github.com/laravel-doctrine/orm/pull/520#discussion_r973195088) in #520

Yeah, to be more precise, the new PrimaryReadReplica was already supported in Laravel Doctrine, but it also supported the old MasterSlaveConnection for backwards compatibility. This has now been completely removed from DBAL 3: https://github.com/doctrine/dbal/blob/3.5.x/UPGRADE.md#removed-masterslaveconnection

**Comment by keithbrink** [comment]
Using this branch within my application and all seems to be good - ready for review.

File: phpunit.xml [hunk 1]
```diff
@@ -1,22 +1,13 @@
...
-         bootstrap="vendor/autoload.php"
-         colors="true"
-         convertErrorsToExceptions="true"
-         convertNoticesToExceptions="true"
-         convertWarningsToExceptions="true"
-         processIsolation="false"
-         stopOnFailure="false"
-    >
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" backupGlobals="false" backupStaticAttributes="false" bootstrap="vendor/autoload.php" colors="true" convertErrorsToExceptions="true" convertNoticesToExceptions="true" convertWarningsToExceptions="true" processIsolation="false" stopOnFailure="false" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
+    <coverage>
```
Comment (eigan) [inline phpunit.xml:14]: What happend here? :)

**Review by eigan — COMMENTED** [review]
Seems to look OK to me, just need to test it a bit 👍🏻

File: phpunit.xml (hunk 1 above)
Comment (keithbrink) [reply phpunit.xml:14]: This was PHPUnit's automatic configuration updater - I'll make it look nice again :-)

**Comment by dpslwk** [comment]
Could you add an `UPGRADE.md` with notes on moving to 2.x version of `laravel-doctrine/orm`

**Comment by keithbrink** [comment]
> Could you add an `UPGRADE.md` with notes on moving to 2.x version of `laravel-doctrine/orm`

Done.

File: UPGRADE.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Upgrade to 2.0
+
+## Breaking Change: DBAL 3
+
+The most significant change in version 2.0 is using doctrine/dbal 3. You should review their upgrade guide in addition to this one: [https://github.com/doctrine/dbal/blob/3.5.x/UPGRADE.md](https://github.com/doctrine/dbal/blob/3.5.x/UPGRADE.md).
```
Comment (eigan) [inline UPGRADE.md]: Tag/branch is removed. Maybe try commit https://github.com/doctrine/dbal/blob/bd54f5043eaff656b314037bf285d8b7f1c311b8/UPGRADE.md (3.6 commit)

File: src/EntityManagerFactory.php [hunk 3]
```diff
@@ -423,15 +423,6 @@ protected function setCustomMappingDriverChain(array $settings, Configuration $c
             'LaravelDoctrine'
         );
 
-        foreach (Arr::get($settings, 'namespaces', []) as $alias => $namespace) {
```
Comment (eigan) [inline src/EntityManagerFactory.php:426]: We can remove this config key now?

Comment (keithbrink) [reply src/EntityManagerFactory.php:426]: Done

File: src/Loggers/Clockwork/DoctrineDataSource.php [hunk 4]
```diff
@@ -59,10 +60,14 @@ protected function getDatabaseQueries()
     {
         $queries = [];
         foreach ($this->logger->queries as $query) {
+            $connection_string = Str::of(
+                (new \ReflectionClass($this->connection->getDatabasePlatform()))->getShortName(),
+            )->lower()->remove('platform')->toString();
```
Comment (eigan) [inline src/Loggers/Clockwork/DoctrineDataSource.php]: Can do this outside the loop

**Review by eigan — COMMENTED** [review]
Mention of `void` cache can be removed from config file.

File: src/Loggers/Clockwork/DoctrineDataSource.php (hunk 4 above)
Comment (keithbrink) [reply src/Loggers/Clockwork/DoctrineDataSource.php]: Done

**Comment by eigan** [comment]
Did some research into the `namespaces` thing.

- We are creating a `MappingDriverChain` with a default driver
- This default driver is using the namespace `LaravelDoctrine` 
  - The driver is registered as default, which means that it can be used even though the namespace doesn't match an entity classname.
  -  The laravel-doctrine/acl package uses this default driver for its own entities (`Permission` entity). It calls the `MappingDriverChain::addPaths()`.
- The namespace support we have in orm `^1` works by reusing the same driver for each namespace

Basically the namespace support doesn't seem to have any effect at all. The entities doesn't need to be in any of the namespaces.

**Comment by eigan** [comment]
@keithbrink I merged 1.7 into 1.8 and noticed the cs-fixer things was already handled there. Sorry about that 🤦🏻‍♂️

**Comment by keithbrink** [comment]
@eigan No problem, I rebased on 1.8.

**Comment by eigan** [comment]
@keithbrink Thank you! This is a good starting point for 2.0. Going to update the other laravel-doctrine packages to allow this version soon. Will try to get this out to our staging env tomorrow, then production next friday.

**Comment by keithbrink** [comment]
@eigan Thanks for getting this across the finish line!
